File: 3-weeks-plan/week2-streaming-airflow/kafka/in_memory_kafka_with_ai.py
import os
import time
import threading
import json
import requests
import pandas as pd
import matplotlib.pyplot as plt
import sqlite3

OLLAMA_HOST = os.getenv("OLLAMA_HOST", "http://localhost:11434")
DB_CONN = sqlite3.connect(':memory:', check_same_thread=False)

# ========== Exchange Data Collectors ==========

def fetch_forex_rates():
    url = "https://api.exchangerate.host/latest?base=USD&symbols=BDT,EUR"
    try:
        response = requests.get(url)
        data = response.json()
        rates = data.get("rates", {})
        return [
            {"market": "Forex", "asset": f"USD/{k}", "price": v, "volume": 1000, "timestamp": int(time.time())}
            for k, v in rates.items()
        ]
    except Exception as e:
        print(f"[Forex] Error: {e}")
        return []

def fetch_crypto_rates():
    url = "https://api.coingecko.com/api/v3/simple/price"
    params = {"ids": "bitcoin,ethereum", "vs_currencies": "usd"}
    try:
        response = requests.get(url, params=params)
        data = response.json()
        results = []
        for coin in ["bitcoin", "ethereum"]:
            if coin in data:
                results.append({
                    "market": "Crypto",
                    "asset": f"{coin.upper()}/USD",
                    "price": data[coin]["usd"],
                    "volume": 1000,
                    "timestamp": int(time.time())
                })
        return results
    except Exception as e:
        print(f"[Crypto] Error: {e}")
        return []

# ========== AI Analysis ==========

def query_ollama(prompt, model="gemma"):
    url = f"{OLLAMA_HOST}/api/generate"
    data = {
        "model": model,
        "prompt": prompt,
        "stream": False
    }
    try:
        response = requests.post(url, json=data)
        response.raise_for_status()
        return response.json().get("response", "")
    except Exception as e:
        print(f"[Ollama] Error: {e}")
        return "AI analysis unavailable due to error."

# ========== In-Memory Kafka Broker ==========

class InMemoryKafkaBroker:
    def __init__(self):
        self.topics = {}
        self.active = False

    def start(self):
        self.active = True

    def stop(self):
        self.active = False

    def create_topic(self, topic):
        if topic not in self.topics:
            self.topics[topic] = []

    def send(self, topic, value):
        if not self.active:
            raise Exception("Broker not started")
        self.create_topic(topic)
        self.topics[topic].append(value)

    def consume(self, topic, timeout=5):
        self.create_topic(topic)
        start = time.time()
        while time.time() - start < timeout:
            if self.topics[topic]:
                return self.topics[topic].pop(0)
            time.sleep(0.1)
        return None

    def is_active(self):
        return self.active

# ========== Producer ==========

def producer(broker, topic, interval=60):
    while broker.is_active():
        forex_trades = fetch_forex_rates()
        for trade in forex_trades:
            broker.send(topic, json.dumps(trade))
            print("[Producer] Produced (forex):", trade)
        crypto_trades = fetch_crypto_rates()
        for trade in crypto_trades:
            broker.send(topic, json.dumps(trade))
            print("[Producer] Produced (crypto):", trade)
        time.sleep(interval)

# ========== Consumer, Storage, Charting, AI ==========

def store_data(trade):
    df = pd.DataFrame([trade])
    df.to_sql('trades', DB_CONN, if_exists='append', index=False)

def load_data():
    return pd.read_sql("SELECT * FROM trades", DB_CONN)

def plot_charts(df):
    if df.empty:
        print("No data to plot.")
        return
    plt.figure(figsize=(8, 4))
    df.groupby('asset')['price'].mean().plot(kind='bar')
    plt.title('Average Price per Asset')
    plt.ylabel('Average Price')
    plt.tight_layout()
    plt.savefig('avg_price_per_asset.png')
    plt.clf()

    df.groupby('asset')['volume'].sum().plot(kind='bar', color='orange')
    plt.title('Total Volume per Asset')
    plt.ylabel('Total Volume')
    plt.tight_layout()
    plt.savefig('total_volume_per_asset.png')
    plt.clf()

def consumer(broker, topic, poll_limit=10):
    i = 0
    while broker.is_active() and i < poll_limit:
        msg = broker.consume(topic)
        if msg:
            trade = json.loads(msg)
            print("[Consumer] Consumed:", trade)
            store_data(trade)
            i += 1
        else:
            time.sleep(1)
    print("[Consumer] Stopping after", i, "messages.")

    # ANALYZE AND PLOT
    df = load_data()
    plot_charts(df)
    if not df.empty:
        prompt = (
            "Given the following market trade data in JSON, provide meaningful insights about current forex and crypto trends. "
            "Consider price patterns, anomalies, or opportunities:\n"
            + df.to_json(orient="records")
        )
        insights = query_ollama(prompt, model="gemma")
        print("\n--- AI Market Insights ---\n", insights)
        print("\nCharts saved as:")
        print("  avg_price_per_asset.png")
        print("  total_volume_per_asset.png")

# ========== MAIN LOGIC ==========

if __name__ == "__main__":
    broker = InMemoryKafkaBroker()
    topic = "money-market"
    broker.start()

    prod_thread = threading.Thread(target=producer, args=(broker, topic, 60), daemon=True)
    cons_thread = threading.Thread(target=consumer, args=(broker, topic, 10), daemon=True)  # 10 messages per run

    prod_thread.start()
    cons_thread.start()

    try:
        input("Press Enter to stop (after first 10 messages and analysis)...\n")
    finally:
        broker.stop()
        prod_thread.join(timeout=2)
        cons_thread.join(timeout=2)
        print("[Main] Shutdown complete.")

File: 3-weeks-plan/week2-streaming-airflow/kafka/inmemory_kafka_with_ai.kt
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import java.sql.Connection
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.Statement
import kotlin.concurrent.thread
import java.time.Instant

val OLLAMA_HOST = System.getenv("OLLAMA_HOST") ?: "http://localhost:11434"
val gson = Gson()
val client = OkHttpClient()
val DB_CONN: Connection = DriverManager.getConnection("jdbc:h2:mem:test;DB_CLOSE_DELAY=-1")

data class Trade(
    val market: String,
    val asset: String,
    val price: Double,
    val volume: Int,
    val timestamp: Long
)

// ========== Exchange Data Collectors ==========

fun fetchForexRates(): List<Trade> {
    val url = "https://api.exchangerate.host/latest?base=USD&symbols=BDT,EUR"
    return try {
        val request = Request.Builder().url(url).build()
        client.newCall(request).execute().use { response ->
            val data = gson.fromJson(response.body?.string(), Map::class.java)
            val rates = data["rates"] as? Map<String, Double> ?: return emptyList()
            rates.map { (k, v) ->
                Trade("Forex", "USD/$k", v, 1000, Instant.now().epochSecond)
            }
        }
    } catch (e: Exception) {
        println("[Forex] Error: $e")
        emptyList()
    }
}

fun fetchCryptoRates(): List<Trade> {
    val url = "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin,ethereum&vs_currencies=usd"
    return try {
        val request = Request.Builder().url(url).build()
        client.newCall(request).execute().use { response ->
            val data = gson.fromJson(response.body?.string(), Map::class.java)
            listOf("bitcoin", "ethereum").mapNotNull { coin ->
                val priceInfo = data[coin] as? Map<*, *>
                val usd = (priceInfo?.get("usd") as? Number)?.toDouble()
                if (usd != null) Trade("Crypto", "${coin.uppercase()}/USD", usd, 1000, Instant.now().epochSecond) else null
            }
        }
    } catch (e: Exception) {
        println("[Crypto] Error: $e")
        emptyList()
    }
}

// ========== AI Analysis ==========

fun queryOllama(prompt: String, model: String = "gemma"): String {
    val url = "$OLLAMA_HOST/api/generate"
    val reqBody = gson.toJson(mapOf("model" to model, "prompt" to prompt, "stream" to false))
    val request = Request.Builder()
        .url(url)
        .post(reqBody.toRequestBody("application/json".toMediaTypeOrNull()))
        .build()
    return try {
        client.newCall(request).execute().use { response ->
            val body = response.body?.string()
            val respJson = gson.fromJson(body, Map::class.java)
            respJson["response"] as? String ?: "AI analysis unavailable due to error."
        }
    } catch (e: Exception) {
        println("[Ollama] Error: $e")
        "AI analysis unavailable due to error."
    }
}

// ========== In-Memory Kafka Broker ==========

class InMemoryKafkaBroker {
    private val topics = mutableMapOf<String, MutableList<String>>()
    @Volatile private var active = false

    fun start() { active = true }
    fun stop() { active = false }
    fun createTopic(topic: String) { topics.putIfAbsent(topic, mutableListOf()) }
    fun send(topic: String, value: String) {
        if (!active) throw Exception("Broker not started")
        createTopic(topic)
        topics[topic]?.add(value)
    }
    fun consume(topic: String, timeout: Long = 5000): String? {
        createTopic(topic)
        val start = System.currentTimeMillis()
        while (System.currentTimeMillis() - start < timeout) {
            topics[topic]?.let { if (it.isNotEmpty()) return it.removeAt(0) }
            Thread.sleep(100)
        }
        return null
    }
    fun isActive(): Boolean = active
}

// ========== Producer ==========

fun producer(broker: InMemoryKafkaBroker, topic: String, interval: Long = 60_000L) {
    while (broker.isActive()) {
        fetchForexRates().forEach { trade ->
            broker.send(topic, gson.toJson(trade))
            println("[Producer] Produced (forex): $trade")
        }
        fetchCryptoRates().forEach { trade ->
            broker.send(topic, gson.toJson(trade))
            println("[Producer] Produced (crypto): $trade")
        }
        Thread.sleep(interval)
    }
}

// ========== Consumer, Storage, Charting, AI ==========

fun storeData(trade: Trade) {
    val stmt = DB_CONN.createStatement()
    stmt.executeUpdate("CREATE TABLE IF NOT EXISTS trades (market VARCHAR, asset VARCHAR, price DOUBLE, volume INT, timestamp BIGINT)")
    val sql = "INSERT INTO trades VALUES ('${trade.market}', '${trade.asset}', ${trade.price}, ${trade.volume}, ${trade.timestamp})"
    stmt.executeUpdate(sql)
    stmt.close()
}

fun loadData(): List<Trade> {
    val stmt = DB_CONN.createStatement()
    val rs: ResultSet = stmt.executeQuery("SELECT * FROM trades")
    val results = mutableListOf<Trade>()
    while (rs.next()) {
        results.add(
            Trade(
                rs.getString("market"),
                rs.getString("asset"),
                rs.getDouble("price"),
                rs.getInt("volume"),
                rs.getLong("timestamp")
            )
        )
    }
    rs.close()
    stmt.close()
    return results
}

// Charting can be implemented using JFreeChart or similar libraries

fun consumer(broker: InMemoryKafkaBroker, topic: String, pollLimit: Int = 10) {
    var i = 0
    while (broker.isActive() && i < pollLimit) {
        val msg = broker.consume(topic)
        if (msg != null) {
            val trade = gson.fromJson(msg, Trade::class.java)
            println("[Consumer] Consumed: $trade")
            storeData(trade)
            i++
        } else {
            Thread.sleep(1000)
        }
    }
    println("[Consumer] Stopping after $i messages.")

    val trades = loadData()
    if (trades.isNotEmpty()) {
        val prompt = """
            Given the following market trade data in JSON, provide meaningful insights about current forex and crypto trends. 
            Consider price patterns, anomalies, or opportunities:
            ${gson.toJson(trades)}
        """.trimIndent()
        val insights = queryOllama(prompt, model = "gemma")
        println("\n--- AI Market Insights ---\n$insights")
        println("\nCharts would be saved here (not implemented in this snippet).")
    }
}

// ========== MAIN LOGIC ==========

fun main() {
    val broker = InMemoryKafkaBroker()
    val topic = "money-market"
    broker.start()

    val prodThread = thread(isDaemon = true) { producer(broker, topic, 60_000L) }
    val consThread = thread(isDaemon = true) { consumer(broker, topic, 10) }

    println("Press Enter to stop (after first 10 messages and analysis)...")
    readLine()
    broker.stop()
    prodThread.join(2000)
    consThread.join(2000)
    println("[Main] Shutdown complete.")
}
